src/main/java/org/rocstreaming/roctoolkit/AutoCloseThread.java

```diff
@@ -0,0 +1,138 @@
+     * Collection of {@link NativeObjectReference} to avoid they get garbage collected.
+     */
+    private final ReferenceCollector<NativeObjectReference> phantomCollector;
+
+    /**
+     * Thread running flag.
+     */
+    private final AtomicBoolean running;
+
+    /**
+     * Create a new <code>AutoCloseThread</code>.
+     */
+    private AutoCloseThread() {
+        setName("AutoCloseThread");
+        referenceQueue = new ReferenceQueue<>();
+        phantomCollector = new ReferenceCollector<>();
+        running = new AtomicBoolean(false);
+    }
+
+    /**
+     * Get <code>AutoCloseThread</code> instance.
+     *
+     * @return the <code>AutoCloseThread</code> singleton instance.
+     */
+    static AutoCloseThread getInstance() {
+        return instance;
+    }
+
+    /**
+     * Get thread running flag.
+     *
+     * @return  true if <code>AutoCloseThread</code> is currently running,
+     *          false otherwise.
+     */
+    boolean isRunning() {
+        return running.get();
+    }
+
+    /**
+     * Add a {@link NativeObject} to <code>AutoCloseThread</code>.
+     *
+     * @param nativeObj     {@link NativeObject} to add.
+     * @param dependsOn     {@link NativeObject} dependency.
+     *
+     * @return              the new {@link NativeObjectReference} associated to the {@link NativeObject}.
+     */
+    NativeObjectReference add(NativeObject nativeObj, NativeObject dependsOn) {
+        NativeObjectReference reference = new NativeObjectReference(nativeObj, dependsOn, referenceQueue);
+        phantomCollector.add(reference);
+        return reference;
+    }
+
+    /**
+     * Add a {@link NativeObject} to <code>AutoCloseThread</code>.
+     *
+     * @param nativeObj     {@link NativeObject} to add.
+     *
+     * @return              the new {@link NativeObjectReference} associated to the {@link NativeObject}.
+     */
+    NativeObjectReference add(NativeObject nativeObj) {
+        return add(nativeObj, null);
+    }
+
+    /**
+     * Remove a reference from <code>AutoCloseThread</code>.
+     *
+     * @param reference     the {@link NativeObjectReference} to remove.
+     */
+    void remove(NativeObjectReference reference) {
+        phantomCollector.remove(reference);
+    }
+
+    /**
+     * Entrypoint method of <code>AutoCloseThread</code>.
+     *
+     * Remove any phantom reachable {@link NativeObjectReference} from the 
+     * {@link AutoCloseThread#referenceQueue} associated with this <code>AutoCloseThread</code> 
+     * and <code>close</code> it.
+     */
+    @Override
+    public void run() {
+        running.set(true);
+        while (isRunning()) {
+            try {
+                NativeObjectReference reference = (NativeObjectReference) referenceQueue.remove(MAX_REMOVE_TIMEOUT_MS);
+                synchronized (this) {
+                    remove(reference);
+                    reference.close();
+                }
+            } catch (Exception e) {
+            }
+        }
```

Comment:
I've supposed that all roc objects must be freed explicitly before shutdown; according to what you said that's not the case, so we can just set `AutoCloseThread` as daemon thread, use blocking `remove()` and get rid of shutdown cleanup.
When the JVM is stopped all process memory (including C heap) is freed and reclaimed by the OS.

Reply:
I can imagine two drawbacks of NOT closing Roc objects at JVM shutdown:

* You have created JVM from C program, run it for a while, then terminated JVM and continued running C program. You'll get a leak in C heap. Even if this is possible, this looks like a very rare case.

* When Roc will implement RTCP, RTSP, and other control protocols, proper closing may perform correct termination, e.g. send BYE or TERADOWN, while just terminating JVM will only close the socket. However the other side anyway should be ready to such situation becasue remote peer can just crash without proper closing.

So it seems that proper closing theoretically may be better in some situations, but is likely not critical. Given that the user can always use AutoCloseable or just call close manually, I think that's not a big deal.

Thus, up to me, if there is an easy way to implement this without hurting performance and complicating code, we can do it. Otherwise, let's use daemon thread.

Reply:
used daemon thread in https://github.com/roc-streaming/roc-java/pull/40